x/coinmaster/bindings/message_plugin.go

```diff
@@ -0,0 +1,132 @@
+package bindings
+
+import (
+	"encoding/json"
+
+	wasmkeeper "github.com/CosmWasm/wasmd/x/wasm/keeper"
+	wasmvmtypes "github.com/CosmWasm/wasmvm/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	coinmasterkeeper "github.com/noria-net/noria/x/coinmaster/keeper"
+	coinmastertypes "github.com/noria-net/noria/x/coinmaster/types"
+)
+
+// CustomMessageDecorator returns decorator for custom CosmWasm bindings messages
+func CustomMessageDecorator(coinmaster *coinmasterkeeper.Keeper) func(wasmkeeper.Messenger) wasmkeeper.Messenger {
+	return func(old wasmkeeper.Messenger) wasmkeeper.Messenger {
+		return &CustomMessenger{
+			wrapped:    old,
+			coinmaster: coinmaster,
+		}
+	}
+}
+
+type CustomMessenger struct {
+	wrapped    wasmkeeper.Messenger
+	coinmaster *coinmasterkeeper.Keeper
+}
+
+var _ wasmkeeper.Messenger = (*CustomMessenger)(nil)
+
+// DispatchMsg executes on the contractMsg.
+func (m *CustomMessenger) DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAddress, contractIBCPortID string, msg wasmvmtypes.CosmosMsg) ([]sdk.Event, [][]byte, error) {
+	if msg.Custom != nil {
+		// only handle the happy path where this is really creating / minting / swapping ...
+		// leave everything else for the wrapped version
+		var contractMsg coinmastertypes.CoinmasterMsg
+		if err := json.Unmarshal(msg.Custom, &contractMsg); err != nil {
+			return nil, nil, sdkerrors.Wrap(err, "coinmaster msg")
+		}
+		if contractMsg.Coinmaster == nil {
+			// return nil, nil, sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "nil coinmaster field")
+			return m.wrapped.DispatchMsg(ctx, contractAddr, contractIBCPortID, msg)
+		}
+		coinmasterMsg := contractMsg.Coinmaster
+
+		if coinmasterMsg.Mint != nil {
+			return m.mint(ctx, contractAddr, coinmasterMsg.Mint)
+		}
+		if coinmasterMsg.Burn != nil {
+			return m.burn(ctx, contractAddr, coinmasterMsg.Burn)
+		}
+	}
+	return m.wrapped.DispatchMsg(ctx, contractAddr, contractIBCPortID, msg)
+}
+
+// mint
+func (m *CustomMessenger) mint(ctx sdk.Context, contractAddr sdk.AccAddress, msg *coinmastertypes.MsgCustomCoinmasterMint) ([]sdk.Event, [][]byte, error) {
+	bz, err := PerformMint(m.coinmaster, ctx, contractAddr, msg)
+	if err != nil {
+		return nil, nil, sdkerrors.Wrap(err, "perform mint denom")
```

Comment:
can the error message be expanded? Might be a bit hard to follow this error message.

Reply:
Absolutely right, improved in 09f84f9a0506252b21569b18c79ecf7d7dea67ea